File: Exercises/2016/01_intro/fem1d/v0.2/fem1d.cpp
#include "fem1d.h"

int main ()
{

  mesh m (a, b, nnodes);
  
  double A[nnodes][nnodes];
  std::fill (&(A[0][0]), &(A[nnodes-1][nnodes]), 0.0);
    
  for (unsigned int iel = 0; iel < m.nels; ++iel)
    {

      double mloc[2][2];
      std::fill (&(mloc[0][0]), &(mloc[1][2]), 0.0);
      
      for (unsigned int inode = 0; inode < 2; ++inode)
        {
          double igrad = (inode == 0 ? 1.0 / m.h : -1.0 / m.h);
          for (unsigned int jnode = 0; jnode < 2; ++jnode)
            {
              double jgrad =  (jnode == 0 ? 1.0 / m.h : -1.0 / m.h);
              mloc[inode][jnode] = igrad * jgrad * m.h;
              A[m.elements[iel][inode]][m.elements[iel][jnode]] += 
                mloc[inode][jnode];
            }
        }
    }

  double f[nnodes];
  std::fill (f, f + nnodes, 0.0);
  
  for (unsigned int iel = 0; iel < m.nels; ++iel)
    {
      double vloc[2];
      std::fill (vloc, vloc + 2, 0.0);
      
      for (unsigned int inode = 0; inode < 2; ++inode)
        {
          vloc[inode] = m.h / 2.0;
          f[m.elements[iel][inode]] += vloc[inode];            
        }
    }

  f[0] = 0;
  f[nnodes - 1] = 0;

  A[0][0] = 1.0;
  std::fill (&(A[0][1]), &(A[0][nnodes]), 0.0);

  A[nnodes-1][nnodes-1] = 1.0;
  std::fill (&(A[nnodes-1][0]), &(A[nnodes-1][nnodes-1]), 0.0);

  double uh[nnodes];
  gauss_seidel (A, f, uh);

  for (unsigned int ii = 0; ii < nnodes; ++ii)
    std::cout << uh[ii] << std:: endl;
      
  return 0;
};

File: Exercises/2016/01_intro/fem1d/fem1d.m
a = 0; 
b = 1;
L = b-a;
nnodes = 50;
nels   = nnodes - 1;
h      = L / nels;
nodes  = [a:h:b];
elements = [1:nnodes-1; 2:nnodes];

%% Initialize the coefficients matrix
A = zeros (nnodes);

%% loop over the intervals to assemble the coefficient matrix:
for iel=1:nels

  %% build local matrix
  mloc = zeros (2);
  
  %% loop over the rows of the local matrix
  for inode =1:2
    
    %% compute gradient of the inode-th test function
    igrad = (-1)^inode / h;

    %% loop over the columns of the local matrix
    for jnode =1:2
      
      %% compute gradient of the jnode-th shape function
      jgrad = (-1)^jnode/h;
      
      %% integrate igrad*jgrad over the element iel and store at mloc (inode, jnode)            
      mloc(inode,jnode) = igrad * jgrad * h;
      
    end 
  end 
  
  %% Assembly:
  for inode =1:2
    for jnode =1:2
      A(elements(inode,iel), elements(jnode,iel)) = ...
      A(elements(inode,iel), elements(jnode,iel)) + ...
      mloc(inode,jnode);
    end
  end
end


%% Initialize the rhs vector
f = zeros (nnodes, 1);

%% loop over the elements to assemble the rhs
for iel = 1:nels

  %% build local vector
  vloc = zeros (2, 1);

  %% loop over the rows of local vector
  for inode =1:2
    
    %% compute the integral of the forcing term
    %% times the inode-th test function integrated
    %% on the interval iel 
    vloc(inode) = h / 2;
    
  end 

  %% Assembly:
  for inode = 1:2
    f(elements(inode, iel)) += ...
    vloc(inode);
  end
end
%% Impose boundary conditions via Iron's trick
f(1)   = 0;
f(end) = 0;

A(1, 1) = 1;
A(1, 2:end) = 0;

A(end, end) = 1;
A(end, 1:end-1) = 0;

%% Solve linear system
uh = A \ f;

%% plot solution
plot (nodes, uh, nodes, nodes .* (1 - nodes) / 2, 'x')
